### PR TITLE
[FEATURE] Implement bar chart panel plugin

### PR DESCRIPTION
**Pull request by ellenwyllie**
Screen Shot 2023-07-18 at 10 36 50 AM
Screen Shot 2023-07-18 at 10 47 08 AM

### Sorting
The data can be sorted in ascending or descending order. By default the bars are sorted in descending order.
Screen Shot 2023-07-18 at 11 22 34 AM

### Percentage Mode
When Percentage is selected from the Mode dropdown, the values are calculated as a percentage. Note that the unit selector is disabled when in Percentage mode.
Screen Shot 2023-07-19 at 10 37 27 AM

Percentage mode is disabled when the unit is a percent:
Screen Shot 2023-07-18 at 11 03 41 AM

# Checklist

- [ ] Pull request has a descriptive title and context useful to a reviewer.
- [ ] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [ ] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [ ] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [ ] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [ ] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

**Comment by eunicorn** [comment]
When there are many series and panel size isn't big enough, it becomes hard to read since the bars and numbers are crowded and overlap each other. Maybe there should there a min height of the bars and it starts scrolling?

image

File: ui/components/src/BarChart/BarChart.tsx [hunk 1]
```diff
@@ -32,14 +33,15 @@ export interface BarChartProps {
   height: number;
   data: BarChartData[] | null;
   unit: UnitOptions;
+  mode: ModeOption;
```
Comment (eunicorn) [inline ui/components/src/BarChart/BarChart.tsx:36]: should this be undefined and make `Value` the default?

File: ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  kind: 'BarChart';
+}
+
+/**
+ * The Options object type supported by the BarChart panel plugin.
+ */
+export interface BarChartOptions {
+  calculation: CalculationType;
+  unit: UnitOptions;
+  sort: SortOption;
```
Comment (eunicorn) [inline ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts:33]: I think sort can also be optional and we can make 'asc' as default

Comment (eunicorn) [reply ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts:33]: In general, i think we want as many defaults as possible so we don't have to explicitly define every option in the json schema

File: ui/panels-plugin/src/plugins/bar-chart/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    label: 'Label 3',
+    value: 0,
+  },
+  {
+    label: 'Label 4',
+    value: 50,
+  },
+];
+
+describe('calculatePercentages', () => {
```
Comment (eunicorn) [inline ui/panels-plugin/src/plugins/bar-chart/utils.test.ts:36]: Yay thanks for adding tests. Maybe also good to add test cases with null values?

File: ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  } = props;
+
+  const { queryResults, isLoading, isFetching } = useDataQueries(); // gets data queries from a context provider, see DataQueriesProvider
+
+  const barData: BarChartData[] = useMemo(() => {
+    if (queryResults[0]?.data === undefined) {
+      return [];
+    }
+    const seriesData: BarChartData[] = [];
+    for (const timeSeries of queryResults[0].data.series) {
```
Comment (eunicorn) [inline ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:39]: Actually i think we need to loop through `queryResults` here since you can have multiple queries, and each query can also have multiple series

Comment (eunicorn) [reply ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:39]: You can find an example using the LegendExample panel in Benchmark dashboard

**Comment by ellenwyllie** [comment]
> When there are many series and panel size isn't big enough, it becomes hard to read since the bars and numbers are crowded and overlap each other. Maybe there should there a min height of the bars and it starts scrolling?

Good point! I just added a min height for the bars: 8b2b30fafca5c028d3438a84ed8dc89eb839b594

https://github.com/perses/perses/assets/22085207/b133b596-758f-404f-b36d-3c279d013a27

File: ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts (hunk 2 above)
Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts]: Makes sense, I updated `unit`, `sort` and `mode` to be optional! ef0a2538ce1f30459e24f92dfd1f01ef2c9207de

File: ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        width={contentDimensions.width}
+        height={contentDimensions.height}
+      >
+        <Skeleton variant="text" width={contentDimensions.width - 20} height={contentDimensions.height / 2} />
+      </Box>
+    );
+  }
+
+  return (
+    <Stack
```
[the remark this answers is not in the document]
Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx]: Removed the `Stack` here 0f1879ea0779d506fd1af602689c184b5cca9203

File: ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+import { BarChart, BarChartData } from '@perses-dev/components';
+import { Box, Stack, Skeleton } from '@mui/material';
+import { useMemo } from 'react';
+import { CalculationsMap } from '@perses-dev/core';
+import { useDataQueries, PanelProps } from '@perses-dev/plugin-system';
+import { BarChartOptions } from './bar-chart-model';
+import { calculatePercentages, sortSeriesData } from './utils';
+
+const PADDING = 24;
```
[the remark this answers is not in the document]
Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx]: Updated! 0f1879ea0779d506fd1af602689c184b5cca9203

File: ui/panels-plugin/src/plugins/bar-chart/utils.test.ts (hunk 3 above)
Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/utils.test.ts:40]: Good idea, updated the tests dfd71b6603b5cca95f0bbe8dbcfa137375d4f47d

File: ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx (hunk 4 above)
Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx]: Updated! f9dc6a03c675261b3259dba007af55aabe3ec478

File: ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts (hunk 2 above)
Comment (eunicorn) [reply ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts]: I'm getting some build errors, I think you also need to update the props in ModeSelector and SortSelector

Comment (ellenwyllie) [reply ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts]: Oops, should be good now

**Comment by eunicorn** [comment]
@ellenwyllie one last thing, there is a merge conflict that you need to resolve

File: ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        height={contentDimensions.height}
+      >
+        <Skeleton variant="text" width={contentDimensions.width - 20} height={contentDimensions.height / 2} />
+      </Box>
+    );
+  }
+
+  return (
+    <Box sx={{ padding: `${PADDING}px` }}>
+      <BarChart
```
Comment (sjcobb) [inline ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:77]: We might need some different options for empty state like in TimeSeriesChart, when adding a brand new BarChart, it looks a little odd how the axisLine shows on the left like this: 

image

At first I wasn't sure what that line was, took a second to remember we default to horizontal orientation with `yAxis.type: 'category'`

File: ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts (hunk 2 above)
Comment (sjcobb) [reply ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts]: Not necessary for this PR, but I like the 'Reset to Defaults' button for this too to make it easy to revert to default spec, it should really be handled by the framework in the future and on StatChart and GaugeChart too
image

File: schemas/panels/bar/bar.cue [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"github.com/perses/perses/schemas/common"
+)
+
+kind: "BarChart"
+spec: close({
+	calculation: common.#calculation
+	unit?:       common.#unit
+	sort?:       "asc" | "desc"
+	mode?:       "value" | "percentage"
```
Comment (sjcobb) [inline schemas/panels/bar/bar.cue:25]: Do we know roughly how this schema maps to Grafana's [bargauge](https://play.grafana.org/d/ktMs4D6Mk/5-bar-charts-and-pie-charts?orgId=1&editPanel=13&showCategory=Value%20options) supported options?

image

File: ui/components/src/BarChart/BarChart.tsx [hunk 9]
```diff
@@ -31,11 +36,12 @@ export interface BarChartProps {
...
+  unit?: UnitOptions;
+  mode?: ModeOption;
 }
 
 export function BarChart(props: BarChartProps) {
-  const { width, height, data, unit } = props;
+  const { width, height, data, unit = { kind: 'Decimal' }, mode = 'value' } = props;
   const chartsTheme = useChartsTheme();
 
   const option: EChartsCoreOption = useMemo(() => {
```
Comment (sjcobb) [inline ui/components/src/BarChart/BarChart.tsx:47]: Seeing some cutoff with longer axis labels, adjusting grid.left or one of the axisLabel options might help

image

Comment (sjcobb) [reply ui/components/src/BarChart/BarChart.tsx:47]: I think we should adjust default to show 'M', 'B' notation on value axis like TimeSeriesChart 

image

Comment (sjcobb) [reply ui/components/src/BarChart/BarChart.tsx:47]: Noticed some cutoff on right side when testing percent mode, I think those [grid](https://echarts.apache.org/en/option.html#grid) properties will help, surprised the theme defaults don't cover us though 😕 

image

**Review by sjcobb — APPROVED** [review]
LGTM! I left some smaller comments but they could just be done in a follow up and I'll get this into v0.38.0 and do a patch release later with the rest

File: schemas/panels/bar/bar.cue (hunk 8 above)
Comment (ellenwyllie) [reply schemas/panels/bar/bar.cue:25]: I don't see a percent mode or anything similar for Grafana's bar gauge. Maybe when we do the conversion we just have to switch it back to the raw values?

**Comment by ellenwyllie** [comment]
> LGTM! I left some smaller comments but they could just be done in a follow up and I'll get this into v0.38.0 and do a patch release later with the rest

@sjcobb I addressed your comments in this PR: https://github.com/perses/perses/pull/1325
